src/modules/samples/cpp/src/so/SampleModule.cpp

```diff
@@ -0,0 +1,395 @@
+            status = EINTR;
+        }
+    }
+    else
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiOpen called with null clientName");
+        status = EINVAL;
+    }
+
+    return handle;
+}
+
+void MmiClose(MMI_HANDLE clientSession)
+{
+    Sample* session = reinterpret_cast<Sample*>(clientSession);
+    if (nullptr != session)
+    {
+        delete session;
+    }
+}
+
+int MmiSet(
+    MMI_HANDLE clientSession,
+    const char* componentName,
+    const char* objectName,
+    const MMI_JSON_STRING payload,
+    const int payloadSizeBytes)
+{
+    int status = MMI_OK;
+    Sample* session = nullptr;
+
+    ScopeGuard sg{[&]()
+    {
+        if (MMI_OK == status)
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiSet(%p, %s, %s, %.*s, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, payload, payloadSizeBytes, status);
+            }
+        }
+        else
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiSet(%p, %s, %s, %.*s, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, payload, payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiSet(%p, %s, %s, -, %d) returned %d", clientSession, componentName, objectName, payloadSizeBytes, status);
+            }
+        }
+    }};
+
+    if (nullptr == clientSession)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiSet called with null clientSession");
+        status = EINVAL;
+        return status;
+    }
+
+    rapidjson::Document document;
+    if (document.Parse(payload).HasParseError())
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiSet unabled to parse JSON payload");
+        status = EINVAL;
+        return status;
+    }
+
+    session = reinterpret_cast<Sample*>(clientSession);
+
+    // Dispatch the request to the appropriate handler for the given component and object
+    if (0 == g_componentName.compare(componentName))
+    {
+        if (0 == g_objectName.compare(objectName))
+        {
+            // Get the required data from the payload and dispatch the request to the client session
+            if (document.IsString())
+            {
+                status = session->SetValue(document.GetString());
+            }
+            else
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiSet JSON payload is not a string");
+                status = EINVAL;
+            }
+        }
+        else
+        {
+            OsConfigLogError(SampleLog::Get(), "MmiSet called with invalid objectName: %s", objectName);
+            status = EINVAL;
+        }
+    }
+    else
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiSet called with invalid componentName: %s", componentName);
+        status = EINVAL;
+    }
```

Comment:
MmiSet should be a simple shell that only invokes the base class Set method.

Same for MmiGet.

See:

[https://github.com/Azure/azure-osconfig/blob/main/docs/modules.md#121-module-static-library](https://github.com/Azure/azure-osconfig/blob/main/docs/modules.md#121-module-static-library)